### PR TITLE
[#48330] Check for authorization state of storages in project storage…

### PR DESCRIPTION
**Pull request by Kharonus**
… create form

- https://community.openproject.org/work_packages/48330
- added storage login button to dynamic components
- splitted new project storage

### CODE REVIEW

File: frontend/src/stimulus/controllers/dynamic/project-storage-form.controller.ts [hunk 1]
```diff
@@ -77,17 +91,24 @@ export default class ProjectStorageFormController extends Controller {
...
-      void fetch(href)
-        .then((data) => data.json())
-        .then((file:IStorageFile) => {
-          this.selectedFolderTextTarget.innerText = file.name;
-        });
-    }
+    combineLatest([
+      this.fetchStorageAuthorizationState(),
+      this.fetchProjectFolder(),
+    ]).subscribe(([isConnected, projectFolder]) => {
```
Comment (akabiru) [inline frontend/src/stimulus/controllers/dynamic/project-storage-form.controller.ts:97]: Q: Curious why we need to wait for the latest values from authorization state and project folder- do we have [_"multiple long-lived observables"_](https://www.learnrxjs.io/learn-rxjs/operators/combination/combinelatest)? 🤔

File: frontend/src/stimulus/controllers/dynamic/project-storage-form.controller.ts [hunk 2]
```diff
@@ -77,17 +91,24 @@ export default class ProjectStorageFormController extends Controller {
...
+        this.selectProjectFolderButtonTarget.style.display = 'inline-block';
+        this.loginButtonTarget.style.display = 'none';
+        this.selectedFolderTextTarget.innerText = projectFolder === null
+          ? this.placeholderFolderNameValue
+          : projectFolder.name;
+      } else {
+        this.selectProjectFolderButtonTarget.style.display = 'none';
+        this.loginButtonTarget.style.display = 'inline-block';
+        this.selectedFolderTextTarget.innerText = this.notLoggedInValidationValue;
+      }
```
Comment (akabiru) [inline frontend/src/stimulus/controllers/dynamic/project-storage-form.controller.ts:108]: Nice one 🎨

File: frontend/src/stimulus/controllers/dynamic/project-storage-form.controller.ts [hunk 3]
```diff
@@ -142,6 +159,33 @@ export default class ProjectStorageFormController extends Controller {
     return `${this.storage._links.self.href}/files/${projectFolderId}`;
   }
 
+  private fetchStorageAuthorizationState():Observable<boolean> {
+    return from(fetch(this.storage._links.self.href)
+      .then((data) => data.json()))
+      .pipe(
+        map((storage:IStorage) => storage._links.authorizationState.href === storageConnected),
```
Comment (akabiru) [inline frontend/src/stimulus/controllers/dynamic/project-storage-form.controller.ts:166]: Ah, I see the reason for `combineLatest` 👍🏾

File: modules/storages/app/views/storages/project_settings/_project_folder_form.html.erb [hunk 4]
```diff
@@ -139,20 +105,29 @@ See COPYRIGHT and LICENSE files for more details.
...
+                                  'project-storage-form-target': "loginButton"
+                                }
+      %>
       <span
         data-project-storage-form-target="selectedFolderText"
-      ></span>
+      >...</span>
     </div>
-    <% if @errors.present? && @errors.any? %>
+    <% if @errors&.any? { |error| error.attribute == :project_folder_id } %>
```
Comment (akabiru) [inline modules/storages/app/views/storages/project_settings/_project_folder_form.html.erb:130]: I'm not sure why we need to filter errors by `project_folder_id`. I would presume we want to show errors generically for the project_storage?

Perhaps a note?

**Review by akabiru — APPROVED** [review]
Nicely done! 👏🏾  _A few clarifying remarks_ 👍🏾

File: frontend/src/stimulus/controllers/dynamic/project-storage-form.controller.ts (hunk 1 above)
Comment (Kharonus) [reply frontend/src/stimulus/controllers/dynamic/project-storage-form.controller.ts:97]: Well, it is true, that `combineLatest` has this use case, but the effect it has is perfect for this use case: it takes an array of observables and fires only if every single of them fired at least once.

`fetch` will return an observable, that fires exactly once. So, if you now read this line, it says:
1. Do 2 API fetch requests
2. Do something with the results ONLY IF both are completed.

File: modules/storages/app/views/storages/project_settings/_project_folder_form.html.erb (hunk 4 above)
Comment (Kharonus) [reply modules/storages/app/views/storages/project_settings/_project_folder_form.html.erb:130]: there is currently only one possible error, but also if there were multiple, this very `span` should be only shown, if this particular error for the attribute `project_folder_id` is among the lists of errors.

File: frontend/src/stimulus/controllers/dynamic/project-storage-form.controller.ts (hunk 1 above)
Comment (akabiru) [reply frontend/src/stimulus/controllers/dynamic/project-storage-form.controller.ts:97]: Nice one, cheers for the explanation 🎓 👍🏾

File: modules/storages/app/views/storages/project_settings/_project_folder_form.html.erb (hunk 4 above)
Comment (akabiru) [reply modules/storages/app/views/storages/project_settings/_project_folder_form.html.erb:130]: Got you, I see we have the `projectFolderIdValidationTarget` for the error setting from the stimulus controller. 👍🏾

**Review by akabiru — COMMENTED** [review]
All good, cheers for the clarifications @Kharonus 🥇

**Review by apfohl — APPROVED** [review]
🦄

**Pull request merged**